File: apps/api/internal/handler/travel_allowance_preview.go
package handler

import (
	"encoding/json"
	"net/http"
	"time"

	"github.com/go-openapi/strfmt"
	"github.com/google/uuid"

	"github.com/tolga/terp/gen/models"
	"github.com/tolga/terp/internal/service"
)

// TravelAllowancePreviewHandler handles travel allowance preview HTTP requests.
type TravelAllowancePreviewHandler struct {
	svc *service.TravelAllowancePreviewService
}

// NewTravelAllowancePreviewHandler creates a new TravelAllowancePreviewHandler.
func NewTravelAllowancePreviewHandler(svc *service.TravelAllowancePreviewService) *TravelAllowancePreviewHandler {
	return &TravelAllowancePreviewHandler{svc: svc}
}

// Preview handles POST /travel-allowance/preview
func (h *TravelAllowancePreviewHandler) Preview(w http.ResponseWriter, r *http.Request) {
	var req models.TravelAllowancePreviewRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "Invalid request body")
		return
	}
	if err := req.Validate(nil); err != nil {
		respondError(w, http.StatusBadRequest, err.Error())
		return
	}

	ruleSetID, err := uuid.Parse(req.RuleSetID.String())
	if err != nil {
		respondError(w, http.StatusBadRequest, "Invalid rule_set_id")
		return
	}

	input := service.TravelAllowancePreviewInput{
		RuleSetID: ruleSetID,
		TripType:  *req.TripType,
	}

	input.DistanceKm = req.DistanceKm
	input.DurationMinutes = int(req.DurationMinutes)

	if !time.Time(req.StartDate).IsZero() {
		input.StartDate = time.Time(req.StartDate)
	}
	if !time.Time(req.EndDate).IsZero() {
		input.EndDate = time.Time(req.EndDate)
	}
	if req.ThreeMonthActive != nil {
		input.ThreeMonthActive = *req.ThreeMonthActive
	}

	result, err := h.svc.Preview(r.Context(), input)
	if err != nil {
		handleTravelPreviewError(w, err)
		return
	}

	respondJSON(w, http.StatusOK, travelPreviewToResponse(result))
}

// travelPreviewToResponse converts internal result to API response.
func travelPreviewToResponse(result *service.TravelAllowancePreviewResult) *models.TravelAllowancePreview {
	taxFreeTotal, _ := result.TaxFreeTotal.Float64()
	taxableTotal, _ := result.TaxableTotal.Float64()
	totalAllowance, _ := result.TotalAllowance.Float64()

	resp := &models.TravelAllowancePreview{
		TripType:       result.TripType,
		RuleSetID:      strfmt.UUID(result.RuleSetID.String()),
		RuleSetName:    result.RuleSetName,
		TaxFreeTotal:   taxFreeTotal,
		TaxableTotal:   taxableTotal,
		TotalAllowance: totalAllowance,
		Breakdown:      make([]*models.TravelAllowanceBreakdownItem, 0, len(result.Breakdown)),
	}

	for _, item := range result.Breakdown {
		taxFreeAmt, _ := item.TaxFreeAmount.Float64()
		taxableAmt, _ := item.TaxableAmount.Float64()
		taxFreeSub, _ := item.TaxFreeSubtotal.Float64()
		taxableSub, _ := item.TaxableSubtotal.Float64()

		resp.Breakdown = append(resp.Breakdown, &models.TravelAllowanceBreakdownItem{
			Description:     item.Description,
			Days:            int64(item.Days),
			TaxFreeAmount:   taxFreeAmt,
			TaxableAmount:   taxableAmt,
			TaxFreeSubtotal: taxFreeSub,
			TaxableSubtotal: taxableSub,
		})
	}

	return resp
}

// handleTravelPreviewError maps service errors to HTTP responses.
func handleTravelPreviewError(w http.ResponseWriter, err error) {
	switch err {
	case service.ErrTravelPreviewRuleSetNotFound:
		respondError(w, http.StatusNotFound, "Rule set not found")
	case service.ErrTravelPreviewRuleSetIDRequired:
		respondError(w, http.StatusBadRequest, "Rule set ID is required")
	case service.ErrTravelPreviewTripTypeRequired:
		respondError(w, http.StatusBadRequest, "Trip type is required")
	case service.ErrTravelPreviewInvalidTripType:
		respondError(w, http.StatusBadRequest, "Trip type must be 'local' or 'extended'")
	case service.ErrTravelPreviewDistanceRequired:
		respondError(w, http.StatusBadRequest, "Distance is required for local travel preview")
	case service.ErrTravelPreviewDurationRequired:
		respondError(w, http.StatusBadRequest, "Duration is required for local travel preview")
	case service.ErrTravelPreviewDatesRequired:
		respondError(w, http.StatusBadRequest, "Start date and end date are required for extended travel preview")
	case service.ErrTravelPreviewNoMatchingRule:
		respondError(w, http.StatusNotFound, "No matching rule found for given distance and duration")
	case service.ErrTravelPreviewNoExtendedRule:
		respondError(w, http.StatusNotFound, "No active extended travel rule found for this rule set")
	default:
		respondError(w, http.StatusInternalServerError, "Internal server error")
	}
}

File: src/trpc/routers/travelAllowancePreview.ts
/**
 * Travel Allowance Preview Router
 *
 * Provides travel allowance calculation preview via tRPC.
 *
 * Replaces the Go backend travel allowance preview endpoint:
 * - POST /travel-allowance-preview -> travelAllowancePreview.preview (query)
 *
 * @see apps/api/internal/service/travel_allowance_preview.go
 * @see apps/api/internal/calculation/travel_allowance.go
 */
import { z } from "zod"
import { createTRPCRouter, tenantProcedure } from "@/trpc/init"
import { requirePermission } from "@/lib/auth/middleware"
import { permissionIdByKey } from "@/lib/auth/permission-catalog"
import { handleServiceError } from "@/trpc/errors"
import * as service from "@/lib/services/travel-allowance-preview-service"

// --- Permission Constants ---

const TRAVEL_ALLOWANCE_MANAGE = permissionIdByKey("travel_allowance.manage")!

// --- Output Schemas ---

const breakdownItemSchema = z.object({
  description: z.string(),
  days: z.number(),
  taxFreeAmount: z.number(),
  taxableAmount: z.number(),
  taxFreeSubtotal: z.number(),
  taxableSubtotal: z.number(),
})

const previewOutputSchema = z.object({
  tripType: z.string(),
  ruleSetId: z.string(),
  ruleSetName: z.string(),
  taxFreeTotal: z.number(),
  taxableTotal: z.number(),
  totalAllowance: z.number(),
  breakdown: z.array(breakdownItemSchema),
})

// --- Input Schemas ---

const previewInputSchema = z.object({
  ruleSetId: z.string(),
  tripType: z.enum(["local", "extended"]),
  distanceKm: z.number().optional().default(0),
  durationMinutes: z.number().int().optional().default(0),
  startDate: z.string().optional(),
  endDate: z.string().optional(),
  threeMonthActive: z.boolean().optional().default(false),
})

// --- Router ---

export const travelAllowancePreviewRouter = createTRPCRouter({
  /**
   * travelAllowancePreview.preview -- Calculates a travel allowance preview.
   *
   * For local trips: matches distance/duration against active local rules.
   * For extended trips: computes day breakdown using active extended rule.
   *
   * Requires: travel_allowance.manage permission
   */
  preview: tenantProcedure
    .use(requirePermission(TRAVEL_ALLOWANCE_MANAGE))
    .input(previewInputSchema)
    .output(previewOutputSchema)
    .query(async ({ ctx, input }) => {
      try {
        return await service.preview(ctx.prisma, ctx.tenantId!, input)
      } catch (err) {
        handleServiceError(err)
      }
    }),
})
